src/ipc/handler.c

```diff
@@ -1231,13 +1232,74 @@ static int ipc_glb_tplg_message(uint32_t header)
 }
 
 #if CONFIG_DEBUG
+
+static int fill_mem_usage_elems(int zone, int elem_number,
+				struct sof_ipc_dbg_mem_usage_elem *elems)
+{
+	int i;
+
+	for (i = 0; i < elem_number; ++i) {
+		elems[i].zone = zone;
+		elems[i].id = i;
+		elems[i].used = rmem_usage(zone, i);
+		elems[i].free = rmem_free(zone, i);
```

Comment:
nitpick: this isn't atomic as it is? So in theory some allocation / freeing can happen between the two calls, in which case `used + free` won't be equal to capacity?

Reply:
good point, so I move `rmem_usage`/`rmem_free` from `alloc.h` to `mm_heap.h`, rename it to `heap_info(struct mm_info *out)` and start using `struct mm_info` as output value instead two different functions .

Reply:
updated